qutip/core/coefficient.py

```
@@ -0,0 +1,621 @@
+
+
+def reduce(coeff, args):
+    """ Reduce decorated string coefficient to 1 object:
+    c = coefficient("t")
+    optimize(c + conj(c)) => coefficient("t+conj(t)")
+    """
+    reduced = coeff.optstr()
+    if reduced:
+        return coefficient(reduced, args=args)
+    else:
+        return coeff
+
+
+# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+# %%%%%%%%%      Everything under this is for string compilation      %%%%%%%%%
+# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+class CompilationOptions:
+    """
+    Options for compilation.
+    use_cython: bool
+        execute strings as python code instead of cython.
+
+    Type management options.
+    accept_int : None, bool
+        Whether integer constants and args are kept or upgraded to float
+        If `None`, use in if array subscrition is used.
+    accept_float : bool
+        Whether float are kept as float or upgraded to complex.
+    no_types : bool
+        Give up on detecting and using c types.
+    recompile : bool
+        Do not use previously made files but build a new one.
+    compiler_flags : str
+        Flags to pass to the compiler, ex: "-Wall -O3"...
+        Flags not matching your comiler and OS may cause compilation to fail.
+        Use "recompile=True", when trying to if the string pattern was
+        previously used.
+    link_flags : str
+        Libraries to link to pass to the compiler. They can not be used to add
+        function to the string coefficient.
+    extra_import : str
+        Cython code to add at the head of the file. Can be used to add extra
+        import or import c code etc. ex:
+        "from scipy.linalg import det"
+        "from qutip.core.data import CSR"
+    """
+    # TODO: use the Options decorator when merged and put in Core options (v5)
+
+    # use cython for compiling string coefficient.
+    try:
+        import cython
+        use_cython = True
+    except:
```

Better to protect this with `except ImportError` - raw `except` can sometimes mask other errors.